### PR TITLE
Prevent endless recursion, limit traversal depth.

### LINKED ISSUE #6674: Loading a Gradle project that uses SpotBugs plugin reports an unknown problem
### Apache NetBeans version

Apache NetBeans 20 release candidate

### What happened

I had a project that was loading on 20-rc3 before,and then after I added teh SpotBugs gradle plugin to my build.gradle file, NB doesn't like it anymore and it won't tell me why.  Resolve project problems shows the empty problem dialog with only a yellow triangle icon with no text in the list of problems.

When I select the "Cannot load:..." notification form the Notifications panel, the side panel shows "Priority: Low,  Category: Warning", and a single bullet point with no text after it.

### How to reproduce

Add the following to a Gradle java-library project:
```groovy
plugins {
    id 'com.github.spotbugs' version '5.2.3'
}

spotbugs {
  // allow throw in a constructor
  omitVisitors = ["ConstructorThrow"] // see https://spotbugs.readthedocs.io/en/latest/detectors.html
}

tasks.withType(com.github.spotbugs.snom.SpotBugsTask) {
  reports {
    xml.enabled false
    html.enabled true
  }
}

tasks.withType(com.github.spotbugs.snom.SpotBugsTask) {
  reports {
    xml.enabled false
    html.enabled true
  }
}
```

### Did this work correctly in an earlier version?

No / Don't know

### Operating System

Ubuntu 23.10

### JDK

17.0.9; OpenJDK 64-Bit Server VM 17.0.9+8-LTS, Amazon Corretto on Intel

### Apache NetBeans packaging

Apache NetBeans provided installer

### Anything else

Happens every time.  Removing the SpotBugs code from `build.gradle` fixes it.

### Are you willing to submit a pull request?

No

### PR DESCRIPTION
**Pull request by sdedic**
Prevents infinite recursion through structures in project model, should fix #6674 and #4779. This is done by:
- in dump of container (iterable, map, named container) avoids to dump a value **equal** to the container itself. This avoids simple recursions, such as with Path reported in #6674
- limits deph of structure traversal
- wraps `toString()` calls in try-catch, so potential error (incl. stack overflow) does not abort and just dumps an object placeholder.

The other part delivers more error details to NB. Fo unexpected exceptions, stack traces are provided in the `detail` Report field. Reports are marked with `severity` so now the loader can declare unexpected (exception) fatal (error), nonfatal (warning) and info messaes.

Unexpected exceptions are presented in project problems with their (truncated to 100 lines) stacktraces. Build system exceptions (i.e. unresolved artifact) should not present stacktraces, but full details from the nested exceptions.

Just error and exception levels pop up project problems window and notifications. Stacktraces are received by NB and can be optionally logged (with appropriate loglevel) - useful for user diagnostic.

I have added a test that a project with Spotbugs plugin loads OK + checks to verify recursive structure avoidance.

### CODE REVIEW

File: extide/gradle/src/org/netbeans/modules/gradle/loaders/LegacyProjectLoader.java [hunk 1]
```diff
@@ -420,10 +454,30 @@ private static GradleReport createReport(Throwable e) {
         } else {
             reported = e;
         }
+        String cn = e.getClass().getName();
+        if (cn.contains("GradleScriptException") || cn.contains("ResolutionException")) {
```
Comment (sdedic) [inline extide/gradle/src/org/netbeans/modules/gradle/loaders/LegacyProjectLoader.java:458]: These are common exception classes for failures reported by the build system itself, mark as 'user' or expected exceptions.

**Review by dbalek — APPROVED** [review]
Looks fine.

**Comment by sdedic** [comment]
@lkishalmi do the changes look sane ? I'd merge it in the following days

File: extide/gradle/src/org/netbeans/modules/gradle/cache/ProjectInfoDiskCache.java
Comment (lkishalmi) [inline extide/gradle/src/org/netbeans/modules/gradle/cache/ProjectInfoDiskCache.java:1]: COMPATIBLE_CACHE_VERSION needs to be increased, in order to force reload the info from Gradle.

**Review by lkishalmi — CHANGES_REQUESTED** [review]
Generally fine, just small but important change is required.

File: extide/gradle/src/org/netbeans/modules/gradle/cache/ProjectInfoDiskCache.java
Comment (sdedic) [reply extide/gradle/src/org/netbeans/modules/gradle/cache/ProjectInfoDiskCache.java:1]: patched into  dd185ef0. Thanks.

**Review by lkishalmi — APPROVED** [review]
Thank you!

**Pull request merged**